src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/MethodSignature.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text;
+
+namespace Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing
+{
+    internal sealed class MethodSignature(MethodInfo method)
+    {
+        private static class Tokens
+        {
+            private static class Internal
+            {
+                public const string Prefix = "<";
+                public const string Postfix = ">";
+            }
+
+            public static class Types
+            {
+                public const char ArityDelimiter = '`';
+            }
+
+            public static class Generics
+            {
+                public const char Start = '<';
+                public const string Separator = ", ";
+                public const char End = '>';
+            }
+
+            public static class Parameters
+            {
+                public static class Names
+                {
+                    public const string ImplicitThis = "this";
+                    public const string Unknown = Internal.Prefix + "unknown" + Internal.Postfix;
+                }
+            }
+        }
+
+        public string ModuleName { get; } = method.Module.Name;
+
+        public string? TypeName { get; } = EmitTypeName(method.DeclaringType);
+
+        public string MethodName { get; } = GetMethodName(method);
```

Comment:
Will this cache the result of `GetMethodName(...)` or will it invoke `GetMethodName` each time someone accesses the property?

Reply:
I can't seem to find explicit documentation, but auto-implementation initialization should be called only once. You can see the backing field initialization with ILSpy if you use C# 5.0 decompilation or look directly at the IL code.